### PR TITLE
feat(slice-2b): workout-history surface with week grouping and e2e (PR7)

### PR DESCRIPTION
**Pull request by leehopper**
## Slice 2b — PR7: frontend workout-history surface + E2E (final PR of the slice)

Implements **Unit 7** of the Slice 2b spec — the week-grouped workout-history surface over the DB-driven query endpoint (PR4), plus the slice's end-to-end proof. This is the last of the 8 stacked PRs for Slice 2b (PR1–PR6b already merged).

### What's included

- **`getWorkoutLogHistory` RTK `infiniteQuery`** (`workout-log.api.ts`) over `POST /api/v1/workouts/logs/query` — opaque keyset cursor as the page param, `getNextPageParam` driven by `nextCursor` (drives "Load older"/`hasNextPage`). Provides the `WorkoutLog` tag so logging a workout refetches the list. It's a read (`query`, not `mutation`), so no antiforgery header — matching the endpoint having no `[RequireAntiforgeryToken]`.
- **ISO-week bucketing helper** — groups the flat, newest-first merged page list into `Week of …` sections client-side. Page-boundary-safe: a week split across two fetched pages renders under a **single** header (sorts defensively by `occurredOn` desc, tiebreak id).
- **Sparse log entry card** — derived distance/duration/pace, a sparse `<dl>` of only the metrics actually present (labels from the shared `WORKOUT_METRIC_META` map, so form + history never drift), the freeform note, and display-only splits as a one-line summary + lazy `<table>` inside a `Collapsible`.
- **`/history` route** reusing the auth + onboarding route guards; a **"Workout history"** link on the home plan view as the entry point.
- Read DTO types (`WorkoutLogDto`, `QueryWorkoutLogs*`, `WorkoutLogSplitDto`) re-exported through the generated barrel (type-only, same pattern as `CreateWorkoutLogResponseDto`).
- Semantic tokens + `motion-reduce:` pairing throughout (DEC-063/DEC-070).

### Proof artifacts (spec § Unit 7)

- **Component tests** — sparse `<dl>` renders only present metrics; week grouping buckets correctly across a page boundary; "Load older" appends/hides; splits one-line summary + lazy table.
- **Playwright E2E** — register/onboard → log a minimum-payload workout → log a rich-payload workout with a note → both appear in the week-grouped history with the rich note + Avg HR. The create + history-query calls hit the **real backend** (onboarding/plan stubbed at the wire, per the slice-2a pattern).

### Verification

- `npm run test` — **453 passed** (47 files; +43 new).
- `npm run build` — tsc + vite clean.
- `npm run lint` — 0 errors.
- `npm run check-contrast` — 28/28 WCAG pairs pass.
- `npx playwright test` — **13/13 e2e pass** against the host-run stack (Postgres + backend on :5001 + Vite).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a dedicated Workout History page (linked from home) showing workouts grouped by ISO week, with formatted distance, duration, pace, date and completion status; shows notes, metrics (e.g., avg HR) and expandable splits; supports pagination with "Load older".

* **Tests**
  * End-to-end tests for workout logging and history display; comprehensive unit tests for history grouping, formatting helpers and history UI components.

### CODE REVIEW

File: frontend/src/app/modules/logging/history/week-grouping.helpers.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ * (tiebreak `workoutLogId` desc, mirroring the backend keyset order) so a
+ * merged multi-page list groups correctly even if the pages interleave.
+ */
+export const groupLogsByIsoWeek = (logs: readonly WorkoutLogDto[]): WorkoutHistoryWeekGroup[] => {
+  const sorted = [...logs].sort((a, b) => {
+    if (a.occurredOn !== b.occurredOn) {
+      return a.occurredOn < b.occurredOn ? 1 : -1
+    }
+    return a.workoutLogId < b.workoutLogId ? 1 : -1
+  })
```
[the remark this answers is not in the document]
Comment (leehopper) [reply frontend/src/app/modules/logging/history/week-grouping.helpers.ts:66]: Done in 3bcefa1 — the comparator now returns `0` when both `occurredOn` and `workoutLogId` are equal. In practice `workoutLogId` is a unique PK, so the equal-key branch only fires comparing an element to itself, but the contract fix is correct. Thanks!

**Review by leehopper — COMMENTED** [review]
### Deep Review

A multi-agent deep review (**Frontier** mode: 7 dimensions → deterministic verification → independent validation → blind challenge) found **no blocking issues**. Correctness, security, cross-file impact, conventions/intent, and type design all came back clean and survived independent scrutiny.

Posting **4 improvement suggestions** as inline comments — 3 test-coverage gaps (🟡) and 1 style simplification (💡). None are blocking.

---
Generated by deep-review | Reviewed up to: 3bcefa173fbcce617dc0022b08771725923e11ce

File: frontend/src/app/modules/logging/history/week-grouping.helpers.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ */
+export const groupLogsByIsoWeek = (logs: readonly WorkoutLogDto[]): WorkoutHistoryWeekGroup[] => {
+  const sorted = [...logs].sort((a, b) => {
+    if (a.occurredOn !== b.occurredOn) {
+      return a.occurredOn < b.occurredOn ? 1 : -1
+    }
+    if (a.workoutLogId === b.workoutLogId) {
+      return 0
+    }
+    return a.workoutLogId < b.workoutLogId ? 1 : -1
```
Comment (leehopper) [inline frontend/src/app/modules/logging/history/week-grouping.helpers.ts:65]: **🟡 [MEDIUM] Week-sort comparator tiebreak + equal-key branches are untested**

The `groupLogsByIsoWeek` comparator has a same-date `workoutLogId`-desc tiebreak and an equal-key `return 0` branch. Every test in `week-grouping.helpers.spec.ts` uses distinct `occurredOn` dates (the `log` helper defaults `workoutLogId` to `occurredOn`), so neither branch is exercised — and the `return 0` branch is exactly what commit `3bcefa1` fixed, shipped without a guarding test. A regression to `return -1` (an antisymmetry violation) or a flipped tiebreak would ship undetected.

**Suggested fix:** Add a test with two logs sharing `occurredOn` but with different `workoutLogId`, asserting `workoutLogId`-desc order within the week; plus a stability test with two entries sharing both keys to cover the `return 0` branch.

_Confidence 80% · blind-challenge 99% · deep-review_

File: frontend/src/app/api/workout-log.api.ts [hunk 3]
```diff
@@ -23,8 +35,35 @@ export const workoutLogApi = apiSlice.injectEndpoints({
...
+        initialPageParam: null,
+        getNextPageParam: (lastPage) => lastPage.nextCursor ?? undefined,
+      },
+      query: ({ pageParam }) => ({
+        url: '/v1/workouts/logs/query',
+        method: 'POST',
+        body: {
+          limit: WORKOUT_HISTORY_PAGE_SIZE,
+          cursor: pageParam,
+        } satisfies QueryWorkoutLogsRequestDto,
```
Comment (leehopper) [inline frontend/src/app/api/workout-log.api.ts:61]: **🟡 [MEDIUM] infiniteQuery pagination contract (getNextPageParam / cursor threading) is untested**

`getWorkoutLogHistory` defines the keyset pagination contract (`initialPageParam: null`, `getNextPageParam: lastPage.nextCursor ?? undefined`, body `{ limit, cursor: pageParam }`), but no test executes it: `history.page.spec.tsx` mocks the hook, and the E2E logs two same-day workouts (one page, null cursor) so “Load older” never fires. `workout-log.api.spec.ts` covers only `createWorkoutLog`. If `getNextPageParam` returned the wrong field or the body dropped `cursor`, “Load older” would loop the first page or stop early — uncaught.

**Suggested fix:** Add a store-level RTK Query test (real `apiSlice` + mocked fetch returning a page with `nextCursor: 'c1'` then a page with `null`) asserting the first request body `cursor` is null, `fetchNextPage` sends `cursor: 'c1'`, and `hasNextPage` flips false on the null-cursor page. Or extend the E2E to span two pages and click “Load older”.

_Confidence 70% · blind-challenge 97% · deep-review_

File: frontend/src/app/modules/logging/history/workout-log-metrics.component.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  metrics?: Record<string, unknown> | null
+}
+
+interface PresentMetric {
+  key: string
+  label: string
+  display: string
+}
+
+const isPresent = (value: unknown): boolean => value !== undefined && value !== null && value !== ''
```
Comment (leehopper) [inline frontend/src/app/modules/logging/history/workout-log-metrics.component.tsx:16]: **🟡 [MEDIUM] Metric value of 0 (a valid boundary) is never tested for presence**

`isPresent` treats numeric `0` as present (`0 !== undefined && 0 !== null && 0 !== ''`), so `elevationGain: 0` or `rpe: 0` should render a row — matching the frontend `CLAUDE.md` rule to use explicit null checks rather than truthiness for `0`. The spec covers non-zero numbers, free text, dropped non-canonical keys, and null/undefined absence, but never a `0` value. A regression to `if (!value) continue` would silently drop `0` metrics with no failing test.

**Suggested fix:** Add a test rendering `WorkoutLogMetrics` with `metrics={{ elevationGain: 0 }}` (and/or `rpe: 0`) asserting the row renders `0 m`, locking in that `0` is treated as present.

_Confidence 65% · blind-challenge 97% · deep-review_

File: frontend/src/app/modules/logging/history/workout-log-splits.component.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+              <tr key={split.index} className="border-b border-border last:border-0">
+                <td className={dataCellClass}>{split.index + 1}</td>
+                <td className={dataCellClass}>{formatDistanceKm(split.distanceMeters) ?? '—'}</td>
+                <td className={dataCellClass}>{formatDuration(split.durationSeconds) ?? '—'}</td>
+                <td className={dataCellClass}>{formatPacePerKm(split.paceSecPerKm) ?? '—/km'}</td>
+                {showHeartRate ? (
+                  <td className={dataCellClass}>
+                    {split.averageHeartRate !== null && split.averageHeartRate !== undefined
+                      ? split.averageHeartRate
+                      : '—'}
```
Comment (leehopper) [inline frontend/src/app/modules/logging/history/workout-log-splits.component.tsx]: **💡 [LOW] HR cell: replace verbose null/undefined ternary with nullish coalescing**

The per-row HR cell guards with `split.averageHeartRate !== null && split.averageHeartRate !== undefined ? split.averageHeartRate : '—'`. Since `averageHeartRate` is `number | null | undefined`, this is behaviorally identical to `split.averageHeartRate ?? '—'` (a `0` bpm still renders `0`) — the form the frontend `CLAUDE.md` prefers and that the sibling Distance/Time/Pace cells already use.

_Confidence 57% · blind-challenge 98% · deep-review_

```suggestion
                    {split.averageHeartRate ?? '—'}
```

File: frontend/src/app/modules/logging/history/week-grouping.helpers.ts (hunk 2 above)
Comment (leehopper) [reply frontend/src/app/modules/logging/history/week-grouping.helpers.ts:65]: Addressed in 7225246. Added two tests to `week-grouping.helpers.spec.ts`:
- **same-date tiebreak** — two logs sharing one `occurredOn` with distinct ids, asserting `workoutLogId`-desc order (`['c-id', 'a-id']`);
- **equal-key `return 0` branch** — two entries sharing both `occurredOn` and `workoutLogId` (distinguished by `distanceMeters`), asserting input order is preserved.

Mutation-verified: reverting the comparator to its pre-`3bcefa1` form makes the equal-key test fail, and flipping the tiebreak direction makes the tiebreak test fail — so neither is tautological.

File: frontend/src/app/api/workout-log.api.ts (hunk 3 above)
Comment (leehopper) [reply frontend/src/app/api/workout-log.api.ts:61]: Addressed in 7225246. Added a store-level RTK Query block in `workout-log.api.spec.ts` that dispatches `getWorkoutLogHistory.initiate` against a stubbed `fetch` (the factory actually runs, unlike the hook-mocked page spec):
- first page sends `{ limit: 20, cursor: null }` (`initialPageParam`);
- a `direction: 'forward'` fetch sends the prior page's `nextCursor` (`cursor: 'cursor-1'`);
- a null `nextCursor` makes `getNextPageParam` return `undefined`, so a forward dispatch issues no further fetch.

Mutation-verified: changing `getNextPageParam` to drop the `nextCursor` makes the test fail.

File: frontend/src/app/modules/logging/history/workout-log-metrics.component.tsx (hunk 4 above)
Comment (leehopper) [reply frontend/src/app/modules/logging/history/workout-log-metrics.component.tsx:16]: Addressed in 7225246. Added a test rendering `WorkoutLogMetrics` with `{ elevationGain: 0, rpe: 0 }`, asserting both `0 m` and `0` render — locking in that `isPresent` treats `0` as present. Mutation-verified: switching `isPresent` to a truthiness check makes the test fail.

File: frontend/src/app/modules/logging/history/workout-log-splits.component.tsx (hunk 5 above)
Comment (leehopper) [reply frontend/src/app/modules/logging/history/workout-log-splits.component.tsx]: Addressed in 1ae00c9 — applied the suggested `?? '—'` simplification. The verbose `!== null && !== undefined` ternary is gone and the HR cell now matches the sibling distance/duration/pace cells.

**Comment by leehopper** [comment]
@coderabbitai full review
